File: grpctest/go/server/calculator_server.go
package server

import (
	"fmt"
	"log"
	"net"

	pb "github.com/dr4ke616/grpctest/proto/calculator"

	"golang.org/x/net/context"
	"google.golang.org/grpc"
)

type Server struct{}

func (s *Server) Evaluate(ctx context.Context, in *pb.ExpressionRequest) (*pb.ExpressionAnswer, error) {
	log.Printf("Handling incoming evaluation request")
	ans := EvaluateExpression(in.Expr)
	return &pb.ExpressionAnswer{Answer: fmt.Sprintf("%d", ans)}, nil
}

func Start(port int) {
	log.Printf("Starting Calculator server on %d...", port)

	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
	if err != nil {
		log.Fatalf("failed to listen: %v", err)
	}

	s := grpc.NewServer()
	pb.RegisterCalculatorServer(s, &Server{})
	s.Serve(lis)
}

File: grpctest/scala/server/CalculatorServer.scala
package server

import java.util.logging.Logger
import javax.script.ScriptEngineManager

import grpctest.proto.calculator.{CalculatorGrpc, ExpressionAnswer, ExpressionRequest}
import io.grpc.{Server, ServerBuilder}

import scala.concurrent.{ExecutionContext, Future}


object CalculatorServer {

  private implicit val log: Logger =
    Logger.getLogger(classOf[CalculatorServer].getName)

  def start(port: Int)(implicit ex: ExecutionContext): CalculatorServer = {
    val server = new CalculatorServer(port)
    server.start()
    server
  }
}

private[server] class CalculatorServer(port: Int)(implicit ex: ExecutionContext, log: Logger) {
  self ⇒

  private[this] var _server: Server = _

  private[server] def start(): Unit = {
    _server = ServerBuilder
      .forPort(port)
      .addService(CalculatorGrpc.bindService(new CalculatorImpl(), ex))
      .build
      .start

    log.info(s"Starting Calculator server on $port...")

    sys.addShutdownHook {
      log.info(s"Shutting down Calculator server...")
      self.stop()
    }
  }

  private def stop(): Unit = if (_server != null) {
    _server.shutdown()
  }

  def blockUntilShutdown(): Unit = if (_server != null) {
    _server.awaitTermination()
  }

  private class CalculatorImpl extends CalculatorGrpc.Calculator {

    private val jsEngine =
      new ScriptEngineManager()
        .getEngineByMimeType("text/javascript")

    override def evaluate(request: ExpressionRequest): Future[ExpressionAnswer] =
      Future.successful {
        log.info("Handling incoming evaluation request")
        ExpressionAnswer(jsEngine.eval(request.expr).toString)
      }
  }
}
